### ISSUE TITLE
Compiler's `addDirective` side effects

### ISSUE THREAD

**Issue by Foxandxss** [angular/angular.js#13057]
Hello, I am one of the maintainers of `angular-ui/bootstrap` and we are doing a massive "refactor" to bring prefixes to our directives.

To not break the current users code, we are just duplicating our directives. So from `foo` we leave it as is and we also create a `prefixedFoo`.

To warn our users that they are using deprecated directives, we are just logging a message out. So far so good.

```javascript
.directive('prefixedFoo', function($prefixedService) {
  return {
    template: '<div class="foo">Foo</div>'
  };
})

.directive('foo', function($service) {
  return {
    template: '<div class="foo">Foo</div>',
    link: function() {
      console.log('foo is deprecated, please use prefixedFoo.');
    }
  };
})
```

What is the problem here? When we add `prefixedFoo` in our app and the compiler process it, it will eventually say:

* Hey, process this node to see if there is any directive (`<div class="foo"></div>`)
* Ok, let me check. Grabs `foo`.
* `addDirective` will inject `foo` successfully and then it will check its restrict.
* Ah, it is not restricted to classes, ignore it.

The problem with that, is that it is going to inject `foo` which is going to load `$service`:

```javascript
.service('$service', function() {
  console.log('$service is deprecated, please use $prefixedService');
})
```

So that deprecation log will be printed out.

So `addDirective` is going to inject `foo` and also load its dependencies.

I know this is not a common case, but took me lots of hours figuring it out and a couple of hacks to bypass this. Just thought that it could be good to bring this to your attention.

Cheers.

PD: A plunker to see it in action http://plnkr.co/edit/UDBP7yBgCw2fD0PapOmj?p=info

**Comment by Narretz** [comment]
You mean the problem is that the deprecated directive injects the deprecated service, so people get a deprecation warning even though they personally didn't inject the service?

**Comment by Narretz** [comment]
So why don't you use the prefixedService in your unprefixedDirective?

**Comment by pkozlowski-opensource** [comment]
> So why don't you use the prefixedService in your unprefixedDirective?

Sure, this would be one way of working around this. But from what I understand this issue was about making sure that a directive doesn't "wake up" if it doesn't match. But I understand that this might be challenging since we need to the actual directive definition object to see if it matches....

**Comment by Foxandxss** [comment]
@pkozlowski-opensource That is what I was telling @wesleycho. Having to wake up the directive to get the DDO.

@Narretz Your solution is a good workaround, but we decided to leave the deprecated directives "in the dark". Every new feat/fix is only applied in the prefixed directive (and services). If we put the deprecated directives to "redirect" to the nonPrefixed ones, people won't find a good reason (other than warnings in the console) to update. So we say: Update ASAP because deprecated directives are not getting any fixes.

At least that is the main idea :P.

My point with this issue was not about discussing what could we do in our library, but more about raising a discussion on `addDirective` behavior.

Cheers :)

**Comment by Narretz** [comment]
Oh, so the problem is that the directive is injected even if it doesn't match. Well, that's something different.

**Comment by Foxandxss** [comment]
@Narretz injected in code, tested agains restrictions and then discarded (because it cannot be applied on classes due to restrict EA). It never reaches the page nor it completely loads. It just wake up, load all its deps and nothing else.

**Comment by gkalpak** [comment]
I'm afraid we can't do anything (that would not be a nasty breaking change). Since the `.directive()` method accepts a factory function, we need to instantiate all services to get to the DDO (note that an injected service might be used in order to construct that DDO in the first place).

Changing the API so that dependencies are not loaded if the directive doesn't match (e.g. having an alternative way to specify the directive's `restrict` property), would be possible, but a potentialy hard-to-debug breking change. At this point, I don't feel it is worth it.

Also, note, that the new `.component()` API _does_ behave that way (no injection, unless the component matches and gets applied to an element).

(Obviously, there are several work-arounds available, such as manually instantiating dependencies using the `$injector`, but I understand that is outside of the scope of this issue.)

--
tl;dr
At this point, I don't think it makes sense to make such a breaking change, considering the small number of usecase that would benefit, the already available APIs and the possible work-arounds.

We could add a backwards-compatible way of specifying the `restrict` without having to invoke the directive factory, but I doubt it would be more helpful than confusing for most people.

I'm going to close this as _won't fix_, but if anyone fancies taking a crack at it, I would be glad to take a look :smiley:

**Comment by Foxandxss** [comment]
That is fine. We did a workaround in the day, and we don't need it anymore. 

Thanks
